designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,956 @@
+1. Allow plugins to specify a `parsers` key so parsers can be referenced the same as rules and processors in plugins.
+
+### The `eslint.config.js` File
+
+The `eslint.config.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports an object: 
+
+```js
+module.exports = {
+    name: "name",
+    files: ["*.js"],
+    ignores: ["*.test.js"],
+    settings: {},
+    languageOptions: {
+        ecmaVersion: 2020,
+        sourceType: "module",
+        globals: {},
+        parser: object || "string",
+        parserOptions: {},
+        linterOptions: {
+            reportUnusedDisableDirectives: "string"
+        }
+    }
+    processor: object || "string",
+    plugins: {}
+    rules: {}
+};
+```
+
+The following keys are new to the `eslint.config.js` format:
+
+* `name` - Specifies the name of the config object. This is helpful for printing out debugging information and, while not required, is recommended for that reason.
+* `files` - Determines the glob file patterns that this configuration applies to.
+* `ignores` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this array of glob patterns are subtracted from the files specified in `files`. If there is no `files` key, then `ignores` acts the same as `ignorePatterns` in `.eslintrc` files.
+
+The following keys are specified the same as in `.eslintrc` files:
+
+* `settings`
+* `rules`
+
+The following keys are specified differently than in `.eslintrc` files:
+
+* `plugins` - an object mapping plugin names to implementations. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
+* `processor` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
+* `languageOptions` - top-level grouping for all options that affect how JavaScript is interpreted by ESLint.
+    * `ecmaVersion` - sets the JavaScript version ESLint should use for these files. This value is copied into `parserOptions` if not already present.
+    * `sourceType` - sets the source type of the JavaScript code to parse. One of `module`, `script`, or `commonjs`.
+    * `parser` - an object or string in `eslint.config.js` files (a string in `.eslintrc`)
+    * `parserOptions` - an object specifying any additional parameters to be passed to the parser.
+    * `linterOptions` - an object for linter-specific settings
+        * `reportUnusedDisableDirectives` - new location for the same option name.
+    * `globals` - any additional global variables to add.
+
+Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `eslint.config.js`, these are all objects or strings (referencing an object in a plugin), requiring users to manually specify the objects to use.
+
+The following keys are invalid in `eslint.config.js`:
+
+* `extends` - replaced by config arrays
```

Comment:
> Probably, we should not use spread properties to extend shareable configs because the shareable configs have different forms: object, array, function. Therefore, extends or something like is still needed.

Yeah, that's a good point. It probably makes sense to add back in `extends` for this purpose.

Reply:
I took a look at specifying this, and I'm not sure `extends` will work. As an example, suppose we have a shareable config called `eslint-config-example` that is defined as this:

```js
module.exports = [
    {
        files: ["*.js"],
        languageOptions: { sourceType: "module" },
        rules: {
            semi: ["error", "always"]
        }
    },
    {
        files: ["*.cjs"],
        languageOptions: { sourceType: "commonjs" },
        rules: {
            semi: ["error", "never"]
        }
    }
];
```

If in my config I do this:

```js
module.exports = [
    {
        files: ["*.js"],
        extends: [ require("eslint-config-example") ]
    }
];
```

If we `&&` the `files` fields, then we'll end up with `sourceType` as `"commonjs"` (because it is last), which is incorrect. I'm not sure the `eslintrc` algorithm for `overrides` makes sense here.

After looking into this some more, I'm thinking we should leave off `extends` for now and see how people end up using this in the wild. We can always add something later to handle this case if it becomes common enough.

In the meantime, we can always recommend that shareable configs should not include `files` or `ignores` if they are meant to be used outside of the current project.

Reply:
> If we && the files fields, then we'll end up with sourceType as "commonjs" (because it is last)

Why? a `*.js` file doesn't match `"*.cjs"`.

Reply:
Oh okay, I got a bit confused by your description. Now I understand, it would only match if the filename matched both `*.js` and `*.cjs` at the same time.

I need to think about this some more. Trying to hack an AND operator into `files` seems a bit messy to me.

It does seem like we need some way of extending `files` patterns from shareable configs, I'm just not sure what we currently do in `overrides` is the right model for this new format.

Reply:
After thinking about this for a bit longer, it seems there are actually two different use cases we need to support:

1. The current `overrides` behavior, which implements an AND between grep patterns. We will need to support this in order to easily allow backwards compatibility with how eslintrc currently works.
2. The case that @kaicataldo was talking about can't actually be solved with an AND, it must be an OR. That is, if a config already has a `files`, then we want to also use that config for a different pattern.

In terms of priority, I think that the AND case is the one that needs solving immediately (I do have an idea for that) for backwards compatibility cases. I think we can defer the OR case for now by recommending that shareable configs designed to be used this way should be a single object without a `files` property, and that would allow us to do patterns like use a spread. Also, because the config file is JavaScript, people will be free to manipulate objects in whatever ways they want, so I think it's less critical for us to provide an out-of-the-box solution from the start.

Reply:
Okay, I've added in a bit about how to accomplish AND in the `files` array.

I think we can successfully solve the OR case with documentation to start, and then we can figure out the patterns people are using most frequently and see if we need to add anything more.